cpp/src/join/join_kernels.cuh

```diff
@@ -382,6 +424,125 @@ __global__ void probe_hash_table(multimap_type multi_map,
   }
 }
 
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief  Performs a nested loop join to find all matching rows
+ between the left and right tables and generate the output for the desired Join operation.
+ *
+ * @param[in] left_table The left table
+ * @param[in] right_table The right table
+ * @param[in] check_row_equality The row equality comparator
+ * @param[in] left_num_rows The number of rows in the left table
+ * @param[in] right_num_rows The number of rows in the right table
+ * @param[out] join_output_l The left result of the join operation
+ * @param[out] join_output_r The right result of the join operation
+ * @param[in,out] current_idx A global counter used by threads to coordinate writes to the global
+ output
+ * @param[in] max_size The maximum size of the output
+ * @param[in] flip_results Whether to flip left and right outputs
+ * @param[in] offset An optional offset
+ * @tparam JoinKind The type of join to be performed
+ * @tparam block_size The number of threads per block for this kernel
+ * @tparam output_cache_size The side of the shared memory buffer to cache join output results
+ *
+ */
+/* ----------------------------------------------------------------------------*/
+template <join_kind JoinKind, cudf::size_type block_size, cudf::size_type output_cache_size>
+__global__ void nested_loop_join(table_device_view left_table,
+                                 table_device_view right_table,
+                                 row_equality check_row_equality,
+                                 const cudf::size_type left_num_rows,
+                                 const cudf::size_type right_num_rows,
```

Comment:
I was trying to match the convention of the hash join code -- on further investigation, there is some subtlety there. The hash join's build/probe convention needs to be able to use a subset of the probe table, which is why the `probe_table_num_rows` argument exists in that function. Since the method is different for nested loop joins, that argument can be removed and replaced with something like `const cudf::size_type left_num_rows = left_table.num_rows();`.